core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1683,8 +1686,48 @@ class KafkaApis(val requestChannel: RequestChannel,
     // If this is considered to leak information about the broker version a workaround is to use SSL
     // with client authentication which is performed at an earlier stage of the connection where the
     // ApiVersionRequest is not available.
+    val apiVersionRequest = request.body[ApiVersionsRequest]
+
+    // flush unofficial client set every hour
+    if (System.currentTimeMillis() - flushStartTime > unofficialClientSetTtl) {
+      unofficialClientIdentities.clear()
```

Comment:
we should not clear the entire cache every hour, instead, use a guava cache with TTL or similar to evict entries that reached TTL.